### PR TITLE
(src update) ENG

### PR DESCRIPTION
**Pull request by K1ngfish3r**
- allnovelfull
- bestlightnovel
- divinedaolibrary
- earlynovel
- freenovelupdates => deleted, reason: [SOLD]
- lightnovelpub
- lightnovelreader => readlitenovel
- lightnovelworld
- lnmtl
- mtlnovel
- mtlreader
- novelbin
- novelforest => deleted, reason: [MIA]
- novelfull
- novelfullme => novelbuddy
- novelhall
- novelsOnline
- novel updates
- scribblehub

### CODE REVIEW

File: plugins/english/lightnovelpub.ts [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+            .join(',');
+
+        return novel;
+    }
+
+    async parsePage(novelPath: string, page: string): Promise<Plugin.SourcePage> {
+        const url = this.site + novelPath + '/chapters/page-' + page;
+        const body = await fetchApi(url).then((res) => res.text());
+        const loadedCheerio = parseHTML(body);
+        await this.sleep(1000);
```
Comment (nyagami) [inline plugins/english/lightnovelpub.ts:94]: again, why do you sleep here?

File: plugins/english/lightnovelworld.ts [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+            .join(',');
+
+        return novel;
+    }
+
+    async parsePage(novelPath: string, page: string): Promise<Plugin.SourcePage> {
+        const url = this.site + novelPath + '/chapters/page-' + page;
+        const body = await fetchApi(url).then((res) => res.text());
+        const loadedCheerio = parseHTML(body);
+        await this.sleep(1000);
```
Comment (nyagami) [inline plugins/english/lightnovelworld.ts:94]: here

File: plugins/english/lnmtl.ts [hunk 3]
```diff
@@ -47,24 +50,30 @@ class LnMTLPlugin implements Plugin.PluginBase {
...
-        const novel: Plugin.SourceNovel = {
-            url,
+        let volumes = JSON.parse(
+            loadedCheerio("main")
+                .next()
+                .html()
+                ?.match(/lnmtl.volumes = \[(.*?)\]/)![0]
+                ?.replace("lnmtl.volumes = ", "") || ""
+        );
+        volumes = volumes.map((volume: { id: number; }) => volume.id);
```
Comment (nyagami) [inline plugins/english/lnmtl.ts:66]:
```suggestion
```

File: plugins/english/lnmtl.ts [hunk 4]
```diff
@@ -86,52 +95,68 @@ class LnMTLPlugin implements Plugin.PluginBase {
...
 
-                const chapterDetails = chapterInfo.data.map(
-                (chapter: { number: number; title: string; slug: string; created_at: string; }) => ({
-                    name: `#${chapter.number} ${chapter.title}`,
-                    url: `${this.site}chapter/${chapter.slug}`,
-                    releaseTime: new Date (chapter.created_at).toISOString(), //converts time obtained to UTC +0, TODO: Make it not convert
-                }));
+            const chapterDetails = chapterInfo.data.map(
+            (chapter: ChapterEntry) => ({
+                page: `Volume ${volumes.number} - ${volumes.title}`,
```
Comment (nyagami) [inline plugins/english/lnmtl.ts:146]:
```suggestion
```
If the novel already had totalPages, adding `page` property for chapter makes no sense

File: plugins/english/LightNovelReader.ts [hunk 5]
```diff
@@ -0,0 +1,189 @@
+import { load as parseHTML } from "cheerio";
+import { isUrlAbsolute } from "@libs/isAbsoluteUrl";
+import { fetchApi, fetchFile } from "@libs/fetch";
+import { Plugin } from "@typings/plugin";
+import { Filters, FilterTypes } from "@libs/filterInputs";
+import dayjs from "dayjs";
+
+class ReadLiteNovel implements Plugin.PluginBase {
+    id = "rln.app";
+    name = "ReadLiteNovel";
```
Comment (nyagami) [inline plugins/english/LightNovelReader.ts:10]: could you rename the file

Comment (K1ngfish3r) [reply plugins/english/LightNovelReader.ts]: done

File: plugins/english/scribblehub.ts [hunk 6]
```diff
@@ -74,23 +74,20 @@ class ScribbleHubPlugin implements Plugin.PluginBase {
         return this.parseNovels(loadedCheerio);
     }
 
-    async parseNovelAndChapters(novelUrl: string): Promise<Plugin.SourceNovel> {
-        const url = novelUrl;
-        const result = await fetchApi(url);
+    async parseNovel(novelPath: string): Promise<Plugin.SourceNovel> {
+        const result = await fetchApi(this.site + novelPath);
```
Comment (nyagami) [inline plugins/english/scribblehub.ts:78]: Can you test this on webtest? I have to set `User-Agent` to undefined to bypass cf

File: plugins/english/scribblehub.ts [hunk 7]
```diff
@@ -149,16 +144,16 @@ class ScribbleHubPlugin implements Plugin.PluginBase {
...
+            const chapterUrl = loadedCheerio(el).find('a').attr('href');
 
             if (!chapterUrl) return;
             chapter.push({
                 name: chapterName,
                 releaseTime: parseISODate(releaseDate),
-                url: chapterUrl,
+                path: chapterUrl.replace(this.site, ''),
             });
         });
```
Comment (nyagami) [inline plugins/english/scribblehub.ts:158]: I see that this source has pages for chatpers, you should implement Plugin.PagePlugin

File: plugins/english/allnovelfull.ts [hunk 8]
```diff
@@ -53,41 +50,35 @@ class AllNovelFullPlugin implements Plugin.PluginBase {
 
         link += `?page=${pageNo}`;
 
-        const headers = new Headers();
-        const body = await fetchApi(link, { headers }).then((result) =>
-            result.text()
-        );
+        const result = await fetch(link);
```
[the remark this answers is not in the document]
Comment (K1ngfish3r) [reply plugins/english/allnovelfull.ts:53]: changed allnovelfull to fetchApi

[the remark this answers is not in the document]
Comment (K1ngfish3r) [reply plugins/english/allnovelfull.ts:53]: added fetchApi

File: plugins/english/earlynovel.ts [hunk 9]
```diff
@@ -16,27 +15,43 @@ class EarlyNovelPlugin implements Plugin.PluginBase {
...
         loadedCheerio(".col-truyen-main > .list-truyen > .row").each(
             (i, el) => {
-                const nUrl = loadedCheerio(el).find("h3.truyen-title > a").attr("href");
-
-                const novelUrl = this.site + nUrl    
+                const novelUrl = loadedCheerio(el).find("h3.truyen-title > a").attr("href");  
                 const novelName = loadedCheerio(el)
                     .find("h3.truyen-title > a")
                     .text();
                 const novelCover = loadedCheerio(el).find("img.cover").attr("src");
```
[the remark this answers is not in the document]
Comment (K1ngfish3r) [reply plugins/english/earlynovel.ts:22]: image

[the remark this answers is not in the document]
Comment (K1ngfish3r) [reply plugins/english/earlynovel.ts:22]: nvm the above, fixed

File: plugins/english/earlynovel.ts [hunk 10]
```diff
@@ -50,33 +65,31 @@ class EarlyNovelPlugin implements Plugin.PluginBase {
...
-        const result = await fetchApi(url, { headers });
+    async parseNovel(novelPath: string): Promise<Plugin.SourceNovel & {totalPages: number}> {
+        const result = await fetchApi(this.site + novelPath);
         const body = await result.text();
 
         let loadedCheerio = parseHTML(body);
+        const lastPageStr = loadedCheerio('a:contains("Last ")')
+            .attr("title")
+            ?.match(/(\d+)/g);
+        const lastPage = Number(lastPageStr?.[1] || "0");
```
[the remark this answers is not in the document]
Comment (K1ngfish3r) [reply plugins/english/earlynovel.ts:82]: fixed

File: plugins/english/earlynovel.ts [hunk 11]
```diff
@@ -2,7 +2,6 @@ import { CheerioAPI, load as parseHTML } from "cheerio";
 import { fetchApi, fetchFile } from "@libs/fetch";
 import { FilterTypes, Filters } from "@libs/filterInputs";
 import { Plugin } from "@typings/plugin";
-import { defaultCover } from "@libs/defaultCover";
 
 class EarlyNovelPlugin implements Plugin.PluginBase {
```
[the remark this answers is not in the document]
Comment (K1ngfish3r) [reply plugins/english/earlynovel.ts:6]: switched

File: plugins/english/lightnovelpub.ts [hunk 12]
```diff
@@ -0,0 +1,246 @@
+import { load as parseHTML } from "cheerio";
+import { fetchApi, fetchFile } from "@libs/fetch";
+import { Plugin } from "@typings/plugin";
+import { Filters, FilterTypes } from "@libs/filterInputs";
+import dayjs from "dayjs";
+
+class LightNovelPub implements Plugin.PluginBase {
```
[the remark this answers is not in the document]
Comment (K1ngfish3r) [reply plugins/english/lightnovelpub.ts:7]: switched

File: plugins/english/lightnovelworld.ts [hunk 13]
```diff
@@ -0,0 +1,246 @@
+import { load as parseHTML } from "cheerio";
+import { fetchApi, fetchFile } from "@libs/fetch";
+import { Plugin } from "@typings/plugin";
+import { Filters, FilterTypes } from "@libs/filterInputs";
+import dayjs from "dayjs";
+
+class LightNovelWorld implements Plugin.PluginBase {
```
[the remark this answers is not in the document]
Comment (K1ngfish3r) [reply plugins/english/lightnovelworld.ts:7]: switched

File: plugins/english/lnmtl.ts [hunk 14]
```diff
@@ -10,12 +10,15 @@ class LnMTLPlugin implements Plugin.PluginBase {
     site = "https://lnmtl.com/";
     version = "1.0.0";
```
[the remark this answers is not in the document]
Comment (K1ngfish3r) [reply plugins/english/lnmtl.ts:12]: switched

File: plugins/english/scribblehub.ts (hunk 6 above)
Comment (K1ngfish3r) [reply plugins/english/scribblehub.ts:78]: yes

File: plugins/english/scribblehub.ts (hunk 7 above)
Comment (K1ngfish3r) [reply plugins/english/scribblehub.ts:158]: image
it uses api to obtain chapters

File: plugins/english/LightNovelReader.ts [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    version = "1.0.0";
+    icon = "src/en/readlitenovel/icon.png";
+    site = "https://rln.app";
+
+    parseAgoDate(date: string){ //parseMadaraDate
```
[the remark this answers is not in the document]
Comment (K1ngfish3r) [reply plugins/english/LightNovelReader.ts]: added undefined

**Pull request merged**